packages/standard-asset-apis/src/RoutedSender.ts

```diff
@@ -0,0 +1,308 @@
+    */
+    createRouteSenderAPI: (route: string, connection: string) => Promise<RouteSenderAPI>;
+
+    /**
+      * This called before the first call to standard:route
+      * can be used to change te storage route for data routing
+      * which is implemented in the sender api
+     */
+    storageRouteHook?: (record: DataEntity) => void|Promise<void>;
+
+    /**
+     * This called after the standard:route is pulled from
+     * can be used to change te storage route for data routing
+     * which is implemented in the sender api
+     */
+    dataRouteHook?: (record: DataEntity) => void|Promise<void>;
+
+    /**
+     * This can be used to track the batches start
+     */
+    batchStartHook?: (batchId: number, route: string, size: number) => void|Promise<void>;
+
+    /**
+     * This can be used to track the batches end
+     */
+    batchEndHook?: (batchId: number, route: string) => void|Promise<void>;
+
+    /**
+     * When a reject is missing the required route or metadata to be processed
+     * this method will be called with the record and error. This must be implemented
+     * because this logic may vary depending on where/how it is being used
+     */
+    rejectRecord: (record: DataEntity, error: unknown) => void;
+
+    constructor(
+        routes: Record<string, string>,
+        options: RoutedSenderOptions
+    ) {
+        this.batchSize = options.batchSize;
+        this.concurrencyAllStorage = options.concurrencyAllStorage ?? Infinity;
+        this.concurrencyPerStorage = options.concurrencyPerStorage ?? 10;
+        this.createRouteSenderAPI = options.createRouteSenderAPI;
+        this.dataRouteHook = options.dataRouteHook;
+        this.storageRouteHook = options.storageRouteHook;
+        this.batchStartHook = options.batchStartHook;
+        this.batchEndHook = options.batchEndHook;
+        this.rejectRecord = options.rejectRecord;
+
+        if (this.batchSize <= 0) {
+            throw new Error(`Expect batch size to be >0, got ${this.batchSize}`);
+        }
+
+        this.events.setMaxListeners(this.batchSize);
+
+        for (const [keyset, connection] of Object.entries(routes)) {
+            const keys = keyset.split(',');
+
+            for (const key of keys) {
+                this.routesDefinitions.set(key.trim(), connection);
+            }
+        }
+
+        if (this.routesDefinitions.has('*') && this.routesDefinitions.has('**')) {
+            throw new Error('routing cannot specify "*" and "**"');
+        }
+    }
+
+    /**
+     * This is called once per route for the lifetime of this instance.
+     * This is used to create multiple sender apis
+    */
+    async initializeRoute(route: string): Promise<void> {
+        if (this.initializedRoutes.has(route)) return;
+
+        if (this.initializingRoutes.has(route)) {
+            logger.debug(`Waiting for sender api to be created for route:${route}`);
+            await once(this.events, route);
```

Comment:
What is the purpose of the event emitter here? It looks like its scheduling an event handler here to go once,but only on the subsequent times the route is called, but wont it also be called in the finally part on line 175 where its executing the first time when that finishes?

From the looks of the code it looks like this is running concurrently with the pMap

Reply:
The event emitter is used with `initializingRoutes` to ensure that the route is only being initialized once since the initialization is async. If it is already being initialized, then it will wait for the initialization to finish before returning so the storage reference to the routeConfig will already be setup.

Reply:
I guess I just don't understand why it needs to wait at all, it could just noop as its being already initialized somewhere else, where this is being called is covered by the pmap so it will wait till its all done anyways, it just seems like a useless wait just to log (potentially)

Reply:
Since multiple calls to `initializeRoute` can happen concurrency, the subsequent calls need to wait for the route to be registered in `initializedRoutes` since once this function is resolve, it is expected to have access to the registered route (mainly for the `batches` and `sender` properties).

Reply:
Yes that is my point, the subsequent calls already wait because its wrapped in the pmap, which does not finish until all the clients are built in the first calls. When the pmap finishes all clients are made and ready, regardless if something finishes fast or slow. Its not that big a deal to hold this back on unnecessary waiting, as the limiting factor is the client creation, it just felt a little off to artificially wait when its already waiting from the pmap

Reply:
They don't wait because of the pMap, if I have two records with the same route, the first one will call `initializeRoute`, since it is async, the second record will be processed and will call `initializeRoute` before the first one finishes. This is why I changed the subsequent calls to wait before resolving since we need to wait for the first call it `initializeRoute` finishes so it can register the route. If it was a serial for loop, then yes this wouldn't happen, or maybe if we sorted the records by route and processed each route serially.